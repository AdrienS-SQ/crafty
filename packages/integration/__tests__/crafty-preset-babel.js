/* global describe, it, expect, jest */

const fs = require("fs");
const path = require("path");

const rimraf = require("rimraf");
const configuration = require("@swissquote/crafty/src/configuration");
const getCommands = require("@swissquote/crafty/src/commands/index");

const testUtils = require("../utils");

const getCrafty = configuration.getCrafty;

// node-forge 0.6.33 doesn't work with jest.
// but selfsigned is fixed on this version
// as we don't use it for now, we can simply mock it
// https://github.com/jfromaniello/selfsigned/issues/16
jest.mock("node-forge");

const PRESET_BABEL = "@swissquote/crafty-preset-babel";

it("Loads crafty-preset-babel and does not register webpack tasks", () => {
  const crafty = getCrafty([PRESET_BABEL], {});

  const loadedPresets = crafty.config.loadedPresets.map(
    preset => preset.presetName
  );
  expect(loadedPresets).toContain(PRESET_BABEL);

  const commands = getCommands(crafty);
  expect(Object.keys(commands)).toContain("jsLint");

  crafty.createTasks();
  expect(Object.keys(crafty.undertaker._registry.tasks())).toEqual([]);
});

it("Loads crafty-preset-babel, crafty-runner-webpack and registers webpack task", () => {
  const config = { js: { myBundle: { source: "css/style.scss" } } };
  const crafty = getCrafty(
    [PRESET_BABEL, "@swissquote/crafty-runner-webpack"],
    config
  );

  const loadedPresets = crafty.config.loadedPresets.map(
    preset => preset.presetName
  );
  expect(loadedPresets).toContain(PRESET_BABEL);
  expect(loadedPresets).toContain("@swissquote/crafty-runner-webpack");

  const commands = getCommands(crafty);
  expect(Object.keys(commands)).toContain("jsLint");

  crafty.createTasks();
  expect(Object.keys(crafty.undertaker._registry.tasks())).toEqual([
    "js_myBundle",
    "js",
    "default"
  ]);
});

it("Fails on double runner with incorrect bundle assignment", () => {
  const config = { js: { myBundle: { source: "css/style.scss" } } };
  const crafty = getCrafty(
    [
      PRESET_BABEL,
      "@swissquote/crafty-runner-gulp",
      "@swissquote/crafty-runner-webpack"
    ],
    config
  );

  const loadedPresets = crafty.config.loadedPresets.map(
    preset => preset.presetName
  );
  expect(loadedPresets).toContain(PRESET_BABEL);
  expect(loadedPresets).toContain("@swissquote/crafty-runner-gulp");
  expect(loadedPresets).toContain("@swissquote/crafty-runner-webpack");

  expect(() => crafty.createTasks()).toThrow(
    "You have multiple runners, please specify a runner for 'myBundle'. Available runners are ['gulp/babel', 'webpack']."
  );
});

it("Fails on double runner with imprecise bundle assignment", () => {
  const config = {
    js: { myBundle: { runner: "gulp", source: "css/style.scss" } }
  };
  const crafty = getCrafty(
    [
      PRESET_BABEL,
      "@swissquote/crafty-preset-typescript",
      "@swissquote/crafty-runner-gulp"
    ],
    config
  );

  const loadedPresets = crafty.config.loadedPresets.map(
    preset => preset.presetName
  );
  expect(loadedPresets).toContain(PRESET_BABEL);
  expect(loadedPresets).toContain("@swissquote/crafty-preset-typescript");
  expect(loadedPresets).toContain("@swissquote/crafty-runner-gulp");

  expect(() => crafty.createTasks()).toThrow(
    "More than one valid runner exists for 'myBundle'. Has to be one of ['gulp/babel', 'gulp/typescript']."
  );
});

it("Fails on non-existing runners", () => {
  const config = {
    js: { myBundle: { runner: "someRunner", source: "css/style.scss" } }
  };
  const crafty = getCrafty(
    [
      PRESET_BABEL,
      "@swissquote/crafty-preset-typescript",
      "@swissquote/crafty-runner-gulp"
    ],
    config
  );

  const loadedPresets = crafty.config.loadedPresets.map(
    preset => preset.presetName
  );
  expect(loadedPresets).toContain(PRESET_BABEL);
  expect(loadedPresets).toContain("@swissquote/crafty-preset-typescript");
  expect(loadedPresets).toContain("@swissquote/crafty-runner-gulp");

  expect(() => crafty.createTasks()).toThrow(
    "Invalid runner 'someRunner' for 'myBundle'. Has to be one of ['gulp/babel', 'gulp/typescript']."
  );
});

it("Assigns bundle only once when runner is specified", () => {
  const config = {
    js: { myBundle: { runner: "webpack", source: "css/style.scss" } }
  };
  const crafty = getCrafty(
    [
      PRESET_BABEL,
      "@swissquote/crafty-runner-gulp",
      "@swissquote/crafty-runner-webpack"
    ],
    config
  );

  const loadedPresets = crafty.config.loadedPresets.map(
    preset => preset.presetName
  );
  expect(loadedPresets).toContain(PRESET_BABEL);
  expect(loadedPresets).toContain("@swissquote/crafty-runner-gulp");
  expect(loadedPresets).toContain("@swissquote/crafty-runner-webpack");

  const commands = getCommands(crafty);
  expect(Object.keys(commands)).toContain("jsLint");

  crafty.createTasks();
  expect(Object.keys(crafty.undertaker._registry.tasks())).toEqual([
    "js_myBundle",
    "js",
    "default"
  ]);
});

it("Lints JavaScript using command", () => {
  process.chdir(path.join(__dirname, "../fixtures/crafty-preset-babel/lints"));
  rimraf.sync("dist");

  const result = testUtils.run(["jsLint", "js/**/*.js"]);

  expect(result).toMatchSnapshot();

  // Files aren't generated on failed lint
  expect(fs.existsSync("dist/js/myBundle.min.js")).toBeFalsy();
  expect(fs.existsSync("dist/js/myBundle.min.js.map")).toBeFalsy();
});

it("Generates IDE Helper", () => {
  process.chdir(path.join(__dirname, "../fixtures/crafty-preset-babel/ide"));
  rimraf.sync(".eslintrc.js");
  rimraf.sync("prettier.config.js");
  rimraf.sync(".gitignore");

  const result = testUtils.run(["ide"]);

  expect(result).toMatchSnapshot();

  expect(testUtils.readForSnapshot(".eslintrc.js")).toMatchSnapshot();

  expect(testUtils.readForSnapshot("prettier.config.js")).toMatchSnapshot();

  expect(testUtils.readForSnapshot(".gitignore")).toMatchSnapshot();
});

it("Lints JavaScript using command, ignore crafty.config.js", () => {
  process.chdir(
    path.join(__dirname, "../fixtures/crafty-preset-babel/lints-ignore-config")
  );
  rimraf.sync("dist");

  const result = testUtils.run([
    "--preset",
    PRESET_BABEL,
    "--ignore-crafty-config",
    "jsLint",
    "crafty.config.js"
  ]);

  expect(result).toMatchSnapshot();

  // Files aren't generated on failed lint
  expect(fs.existsSync("dist/js/myBundle.min.js")).toBeFalsy();
  expect(fs.existsSync("dist/js/myBundle.min.js.map")).toBeFalsy();
});

it("Lints JavaScript using command, legacy", () => {
  process.chdir(
    path.join(__dirname, "../fixtures/crafty-preset-babel/lints-es5")
  );
  rimraf.sync("dist");

  const result = testUtils.run(["jsLint", "js/**/*.js", "--preset", "legacy"]);

  expect(result).toMatchSnapshot();

  // Files aren't generated on failed lint
  expect(fs.existsSync("dist/js/myBundle.min.js")).toBeFalsy();
  expect(fs.existsSync("dist/js/myBundle.min.js.map")).toBeFalsy();
});

it("Lints JavaScript using command, recommended preset", () => {
  process.chdir(path.join(__dirname, "../fixtures/crafty-preset-babel/lints"));
  rimraf.sync("dist");

  const result = testUtils.run([
    "jsLint",
    "js/**/*.js",
    "--preset",
    "recommended"
  ]);

  expect(result).toMatchSnapshot();

  // Files aren't generated on failed lint
  expect(fs.existsSync("dist/js/myBundle.min.js")).toBeFalsy();
  expect(fs.existsSync("dist/js/myBundle.min.js.map")).toBeFalsy();
});

it("Lints JavaScript using command, explicit configuration", () => {
  process.chdir(path.join(__dirname, "../fixtures/crafty-preset-babel/lints"));
  rimraf.sync("dist");

  const result = testUtils.run([
    "jsLint",
    "js/**/*.js",
    "--config",
    "eslintOverride.json"
  ]);

  expect(result).toMatchSnapshot();

  // Files aren't generated on failed lint
  expect(fs.existsSync("dist/js/myBundle.min.js")).toBeFalsy();
  expect(fs.existsSync("dist/js/myBundle.min.js.map")).toBeFalsy();
});
