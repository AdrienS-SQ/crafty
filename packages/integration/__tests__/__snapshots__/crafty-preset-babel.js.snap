// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generates IDE Helper 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "
[__:__:__] Starting Crafty __version__...
Written .eslintrc.js
",
}
`;

exports[`Generates IDE Helper 2`] = `
"// This configuration was generated by Crafty
// This file is generated to improve IDE Integration
// You don't need to commit this file, nor need it to run \`crafty build\`

module.exports = {
    \\"plugins\\": [
        \\"@swissquote/swissquote\\"
    ],
    \\"extends\\": [
        \\"plugin:@swissquote/swissquote/format\\"
    ],
    \\"rules\\": {
        \\"no-var\\": 0
    },
    \\"settings\\": {}
};
"
`;

exports[`Lints JavaScript using command 1`] = `
Object {
  "status": 1,
  "stderr": "",
  "stdout": "
[__:__:__] Starting Crafty __version__...

__PATH__/packages/integration/fixtures/crafty-preset-babel/lints/js/Component.js
  11:3   error  Opening curly brace does not appear on the same line as controlling statement  brace-style
  12:13  error  Multiple spaces found before '<'                                               no-multi-spaces

__PATH__/packages/integration/fixtures/crafty-preset-babel/lints/js/script.js
   3:24  error    Multiple spaces found before '\\"./Component.js\\"'                                no-multi-spaces
   5:10  error    'test' is defined but never used                                               no-unused-vars
   5:19  error    A space is required after ','                                                  comma-spacing
   6:1   error    Expected indentation of 2 spaces but found 3                                   indent
  11:5   warning  Unexpected console statement                                                   no-console
  15:5   warning  Unexpected console statement                                                   no-console
  15:19  error    Multiple spaces found before '\\"some method\\"'                                   no-multi-spaces
  15:33  error    There should be no space before ','                                            comma-spacing
  19:3   error    Opening curly brace does not appear on the same line as controlling statement  brace-style
  20:14  error    Expected '===' and instead saw '=='                                            eqeqeq

✖ 12 problems (10 errors, 2 warnings)
  8 errors and 0 warnings potentially fixable with the \`--fix\` option.

",
}
`;

exports[`Lints JavaScript using command, explicit configuration 1`] = `
Object {
  "status": 1,
  "stderr": "",
  "stdout": "
[__:__:__] Starting Crafty __version__...

__PATH__/packages/integration/fixtures/crafty-preset-babel/lints/js/Component.js
  11:3  error  Opening curly brace does not appear on the same line as controlling statement  brace-style

__PATH__/packages/integration/fixtures/crafty-preset-babel/lints/js/script.js
   5:10  error    'test' is defined but never used                                               no-unused-vars
   6:1   error    Expected indentation of 2 spaces but found 3                                   indent
  11:5   warning  Unexpected console statement                                                   no-console
  15:5   warning  Unexpected console statement                                                   no-console
  19:3   error    Opening curly brace does not appear on the same line as controlling statement  brace-style
  20:14  error    Expected '===' and instead saw '=='                                            eqeqeq

✖ 7 problems (5 errors, 2 warnings)
  3 errors and 0 warnings potentially fixable with the \`--fix\` option.

",
}
`;

exports[`Lints JavaScript using command, ignore crafty.config.js 1`] = `
Object {
  "status": 1,
  "stderr": "",
  "stdout": "
[__:__:__] Starting Crafty __version__...

__PATH__/packages/integration/fixtures/crafty-preset-babel/lints-ignore-config/crafty.config.js
  1:1  error  'module' is not defined  no-undef

✖ 1 problem (1 error, 0 warnings)

",
}
`;

exports[`Lints JavaScript using command, legacy 1`] = `
Object {
  "status": 1,
  "stderr": "",
  "stdout": "
[__:__:__] Starting Crafty __version__...

__PATH__/packages/integration/fixtures/crafty-preset-babel/lints-es5/js/script.js
  3:1   error  Use the function form of 'use strict'                                          strict
  3:19  error  A space is required after ','                                                  comma-spacing
  4:1   error  Expected indentation of 2 spaces but found 3                                   indent
  7:1   error  Use the function form of 'use strict'                                          strict
  8:1   error  Opening curly brace does not appear on the same line as controlling statement  brace-style
  9:12  error  Expected '===' and instead saw '=='                                            eqeqeq

✖ 6 problems (6 errors, 0 warnings)
  3 errors and 0 warnings potentially fixable with the \`--fix\` option.

",
}
`;

exports[`Lints JavaScript using command, recommended preset 1`] = `
Object {
  "status": 1,
  "stderr": "",
  "stdout": "
[__:__:__] Starting Crafty __version__...

__PATH__/packages/integration/fixtures/crafty-preset-babel/lints/js/Component.js
  11:3   error  Opening curly brace does not appear on the same line as controlling statement  brace-style
  12:13  error  Multiple spaces found before '<'                                               no-multi-spaces

__PATH__/packages/integration/fixtures/crafty-preset-babel/lints/js/script.js
   3:24  error    Multiple spaces found before '\\"./Component.js\\"'                                no-multi-spaces
   5:10  error    'test' is defined but never used                                               no-unused-vars
   5:19  error    A space is required after ','                                                  comma-spacing
   6:1   error    Expected indentation of 2 spaces but found 3                                   indent
  11:5   warning  Unexpected console statement                                                   no-console
  15:5   warning  Unexpected console statement                                                   no-console
  15:19  error    Multiple spaces found before '\\"some method\\"'                                   no-multi-spaces
  15:33  error    There should be no space before ','                                            comma-spacing
  19:3   error    Opening curly brace does not appear on the same line as controlling statement  brace-style
  20:14  error    Expected '===' and instead saw '=='                                            eqeqeq

✖ 12 problems (10 errors, 2 warnings)
  8 errors and 0 warnings potentially fixable with the \`--fix\` option.

",
}
`;
